src/cdk/portal/portal.ts

```diff
@@ -103,42 +104,38 @@ export class ComponentPortal<T> extends Portal<ComponentRef<T>> {
   }
 }
 
-
 /**
  * A `TemplatePortal` is a portal that represents some embedded template (TemplateRef).
  */
-export class TemplatePortal extends Portal<Map<string, any>> {
+export class TemplatePortal<C> extends Portal<C> {
   /** The embedded template that will be used to instantiate an embedded View in the host. */
-  templateRef: TemplateRef<any>;
+  templateRef: TemplateRef<C>;
 
   /** Reference to the ViewContainer into which the template will be stamped out. */
   viewContainerRef: ViewContainerRef;
 
-  /**
-   * Additional locals for the instantiated embedded view.
-   * These locals can be seen as "exports" for the template, such as how ngFor has
-   * index / event / odd.
-   * See https://angular.io/docs/ts/latest/api/core/EmbeddedViewRef-class.html
-   */
-  locals: Map<string, any> = new Map<string, any>();
+  context: C | undefined;
 
-  constructor(template: TemplateRef<any>, viewContainerRef: ViewContainerRef) {
+  constructor(template: TemplateRef<any>, viewContainerRef: ViewContainerRef, context?: C) {
```

Comment:
This is required to support the abstraction in [dialog.ts](https://github.com/shlomiassaf/material2/blob/9ee6478063fd081b824d206abc93ebe803947d43/src/lib/dialog/dialog.ts#L232-L233)

If you follow the trail, the `attach` method is never called and `BasePortalHost` (`PortalHostDirective) handles it from the top.

Adding a context to the method call of `attachTemplatePortal` is a mess since it involves changing a lot of classes that extend `BasePortal`, for example `MdDialogContainer`

Additionally, changing the signature of `attachTemplatePortal` to accept a context will make it different from the signature `attachComponentPortal`...